### ISSUE TITLE
SetDefaults on connections with wrong DataType crashes Nest

### ISSUE THREAD

**Issue by flinz** [nest/nest-simulator#545]
In the [master branch](https://github.com/nest/nest-simulator/commit/1fd378f8d95d64ed0af32fadd373a9500a68e719) calling ``SetDefaults`` on a connection with a wrong datatype causes Nest to ``terminate called without an active exception``.  In Nest 2.10 this correctly raised a NestError. Importantly, for neuron models this still works correctly, see below. 

Missing we be missing something in our unittests?
I am unsure whether this is connected to #249, since this actually causes a crash not a silent fail.

Test Bash Script
```bash
echo "Try setting neuron C_m with float"
nest -c "/iaf_neuron << /C_m 10. >> SetDefaults"

echo "------"
echo "Try setting synapse tau_plus with float"
nest -c "/stdp_synapse << /tau_plus 10. >> SetDefaults"

echo "------"
echo "Try setting neuron C_m with int (should fail)"
nest -c "/iaf_neuron << /C_m 10 >> SetDefaults"

echo "------"
echo "Try setting synapse tau_plus with int (should fail)"
nest -c "/stdp_synapse << /tau_plus 10 >> SetDefaults"
```

In [Nest 2.10](https://github.com/nest/nest-simulator/releases) this behaves as expected
```bash
Try setting neuron C_m with float
------
Try setting synapse tau_plus with float
------
Try setting neuron C_m with int (should fail)

Nov 11 16:13:39 SetDefaults_l_D [Error]: TypeMismatch
    Expected datatype: doubletype
    Provided datatype: integertype
------
Try setting synapse tau_plus with int (should fail)

Nov 11 16:13:39 SetDefaults_l_D [Error]: TypeMismatch
    Expected datatype: doubletype
    Provided datatype: integertype
```

In the [master branch](https://github.com/nest/nest-simulator/commit/1fd378f8d95d64ed0af32fadd373a9500a68e719) the last command crashes Nest
```bash
Try setting neuron C_m with float
------
Try setting synapse tau_plus with float
------
Try setting neuron C_m with int (should fail)

Nov 11 16:23:52 SetDefaults_l_D [Error]: TypeMismatch
    Expected datatype: doubletype
    Provided datatype: integertype
------
Try setting synapse tau_plus with int (should fail)

terminate called without an active exception
test.sli: line 14: 41387 Abort trap: 6           nest -c "/stdp_synapse << /tau_plus 10 >> SetDefaults"
```

**Comment by heplesser** [comment]
The reason is most likely that we have parallelized setting of defaults across threads [see here](https://github.com/nest/nest-simulator/blob/ad9365396b85edab65a9d2cccfaebdce2a42d477/nestkernel/model_manager.cpp#L361) and that this code does not handle exceptions safely inside the parallel region. This may be because the code only catches `BadProperty`, no other exceptions. The code also has only a single pointer to store the exception for all threads, which may lead to contention if two threads raise an exception simultaneously.

For safe handling of exception in a threaded context, see `ConnBuilder`: [catch and store](https://github.com/nest/nest-simulator/blob/ad9365396b85edab65a9d2cccfaebdce2a42d477/nestkernel/conn_builder.cpp#L754) and [re-throw in serial context](https://github.com/nest/nest-simulator/blob/ad9365396b85edab65a9d2cccfaebdce2a42d477/nestkernel/conn_builder.cpp#L447).

Furthermore, I do not quite understand why most of the threaded context is [marked as critical](https://github.com/nest/nest-simulator/blob/ad9365396b85edab65a9d2cccfaebdce2a42d477/nestkernel/model_manager.cpp#L368) --- this just serializes everything.

@janhahne Could you take a look?

**Comment by flinz** [comment]
This is resolved by #704 .
